Add save integrity levels.

darksylinc opened the pull request:
This PR adds an argument to most save functions to specify what level of integrity should Godot attempt to guarantee when writing a file.

This in preparation for PR #98361; which will introduce a full sync barrier when performing save operations to avoid data loss.

This PR puts such expensive sync() behind the enum SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC; which should be used sparingly for important information.

This PR also sets most cache files to be saved as either SAVE_INTEGRITY_SAVE_SWAP (old default behavior) or SAVE_INTEGRITY_NONE; because Godot can easily generate thousands of these file saving operations. Performing thousands of full syncs() would be too expensive, specially when these files can be regenerated if they're corrupted or lost.

The PR is incomplete as it does not yet expose integrity levels to the user, which is relevant for certain file saving operations, such as when game save files that store the player's progress (those should be done with SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC).

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

darksylinc commented:
The function `ResourceFormatSaver::save` performs a GDVIRTUAL_CALL():

```cpp
Error ResourceFormatSaver::save(const Ref<Resource> &p_resource, const String &p_path, uint32_t p_flags, const int p_integrity_level) {
	Error err = ERR_METHOD_NOT_FOUND;
	GDVIRTUAL_CALL(_save, p_resource, p_path, p_flags, p_integrity_level, err);
	return err;
}
```

I don't know if this breaks the GD ABI and if so, how to implement a fallback function.

bruvzg commented on core/io/file_access.h:
```diff
@@ -247,8 +294,17 @@ class FileAccess : public RefCounted {
+		WARN_DEPRECATED_MSG("Use set_default_integrity_level() instead.");
+		set_default_integrity_level(p_enable ? SAVE_INTEGRITY_SAVE_SWAP : SAVE_INTEGRITY_NONE);
+	}
+	static bool is_backup_save_enabled() {
+		WARN_DEPRECATED_MSG("Use get_default_integrity_level() instead.");
+		return default_save_integrity_level > SAVE_INTEGRITY_NONE;
+	}
```
`is_backup_save_enabled` can be removed entirely, it was only used in `FileAccess` implementation and not exposed.

bruvzg commented on core/io/file_access.h:
```diff
@@ -247,8 +294,17 @@ class FileAccess : public RefCounted {
 
-	static void set_backup_save(bool p_enable) { backup_save = p_enable; }
-	static bool is_backup_save_enabled() { return backup_save; }
+	static void set_backup_save(bool p_enable) {
+		WARN_DEPRECATED_MSG("Use set_default_integrity_level() instead.");
+		set_default_integrity_level(p_enable ? SAVE_INTEGRITY_SAVE_SWAP : SAVE_INTEGRITY_NONE);
+	}
```
```suggestion
#ifndef DISABLE_DEPRECATED
	static void set_backup_save(bool p_enable) {
		WARN_DEPRECATED_MSG("Use set_default_integrity_level() instead.");
		set_default_integrity_level(p_enable ? SAVE_INTEGRITY_SAVE_SWAP : SAVE_INTEGRITY_NONE);
	}
#endif
```

Also, the same for the `OS::set_use_file_access_save_and_swap` in core bind.

bruvzg commented on core/io/file_access.cpp:
```diff
@@ -821,6 +828,11 @@ String FileAccess::get_file_as_string(const String &p_path, Error *r_error) {
 	return ret;
 }
 
+void FileAccess::set_default_integrity_level(SaveIntegrityLevel p_integrity_level) {
```
New method (and corresponding constants) should be probably exposed to scripting in `FileAccess::_bind_methods()`.

Also, `_open` wrapper used for scripting should be updated (and have compatibility method added) to have a new `SaveIntegrityLevel p_integrity_level` argument.

RedMser replied:
Also when adding this binding, you will also want to expose the `SaveIntegrityLevel` enum. The documentation on the enum can then be moved out of the code comments into the class reference .xml files.

bruvzg reviewed (commented):
There's also `FileAccess::set_backup_save(EDITOR_GET("filesystem/on_save/safe_save_on_backup_then_rename"));` call in the `EditorNode` and editor setting that it's using, both should be updated to use the new integrity level.

RandomShaper commented on core/io/resource_format_binary.cpp:
```diff
@@ -2142,7 +2142,7 @@ static String _resource_get_class(Ref<Resource> p_resource) {
 }
 
-Error ResourceFormatSaverBinaryInstance::save(const String &p_path, const Ref<Resource> &p_resource, uint32_t p_flags) {
+Error ResourceFormatSaverBinaryInstance::save(const String &p_path, const Ref<Resource> &p_resource, uint32_t p_flags, const int p_integrity_level) {
```
```suggestion
Error ResourceFormatSaverBinaryInstance::save(const String &p_path, const Ref<Resource> &p_resource, uint32_t p_flags, int p_integrity_level) {
```

I see you're adding `const` to some `int` parameters but not to others. My take is the same as always, especially for parameters, it makes readability worse. In any case, if there was consensus to use `const` everywhere it fits, let's do a whole-codebase pass. I'd therefore suggest not adding them in this PR.

RandomShaper commented:
It may be better to fuse the integrity flags with the mode flags to preserve compatibility.

darksylinc commented:
> It may be better to fuse the integrity flags with the mode flags to preserve compatibility.

Ahhh!!!! That's a lot more KISS. I like it. Way less effort and more backwards compatible. I'll incorporate that change.

darksylinc commented:
> > It may be better to fuse the integrity flags with the mode flags to preserve compatibility.
> 
> Not sure if it's better compatibility:
> 
>     * custom `FileAccess` implementations might use (internal do exactly this) `(mode_flags == WRITE)` or similar checks.
> 
>     * `ModeFlags` name imply that it is i/o mode of operation not generic flags, and it is defined as `Enum`, not `BitField` (renaming it and changing type is more breaking than adding new argument).

Gotcha. Yeah, I think I prefer the current solution of using a separate enum.

Slightly unrelated to that, I think I prefer that, instead of modifying `ResourceSaver::save` to add `p_integrity_level` to each function overload, to modify SaverFlags:

```cpp
enum SaverFlags {
		FLAG_NONE = 0,
		FLAG_RELATIVE_PATHS = 1,
		FLAG_BUNDLE_RESOURCES = 2,
		FLAG_CHANGE_PATH = 4,
		FLAG_OMIT_EDITOR_PROPERTIES = 8,
		FLAG_SAVE_BIG_ENDIAN = 16,
		FLAG_COMPRESS = 32,
		FLAG_REPLACE_SUBRESOURCE_PATHS = 64,
		FLAG_SAVE_INTEGRITY_NONE = 128, // New flag!
		FLAG_SAVE_INTEGRITY_SAVE_SWAP = 256, // New flag!
		FLAG_SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC = FLAG_SAVE_INTEGRITY_NONE | FLAG_SAVE_INTEGRITY_SAVE_SWAP, // New flag!
	};
```

Not setting any flag will use `SAVE_INTEGRITY_DEFAULT`. Else you can explicitly specify what save integrity you want. And this is a lot more backwards compatible.

RandomShaper commented:
As much as I'm a fan of trying to overload the current enums, that looks too awkward:
`swap + sync = none + swap`

darksylinc commented:
I've got a question.

There is this validation error:

```
Validate extension JSON: Error: Field 'classes/ResourceFormatSaver/methods/_save/arguments': size changed value in new API, from 3 to 4.
```

That belongs to a GDVIRTUAL:

```cpp
class ResourceFormatSaver : public RefCounted {
	// BEFORE:
	GDVIRTUAL3R(Error, _save, Ref<Resource>, String, uint32_t)
	// AFTER:
	GDVIRTUAL4R(Error, _save, Ref<Resource>, String, uint32_t, int)
};

Error ResourceFormatSaver::save(const Ref<Resource> &p_resource, const String &p_path, uint32_t p_flags, int p_integrity_level) {
	Error err = ERR_METHOD_NOT_FOUND;
	// BEFORE:
	GDVIRTUAL_CALL(_save, p_resource, p_path, p_flags, err);
	// AFTER:
	GDVIRTUAL_CALL(_save, p_resource, p_path, p_flags, p_integrity_level, err);
	return err;
}

void ResourceFormatSaver::_bind_methods() {
	// BEFORE:
	GDVIRTUAL_BIND(_save, "resource", "path", "flags");
	// AFTER:
	GDVIRTUAL_BIND(_save, "resource", "path", "flags", "integrity_level");
}
```

Is there a way to deal with this in a backwards compatible way? Or do we just drop the integrity_level argument?

dsnopek commented:
> Is there a way to deal with this in a backwards compatible way?

Unfortunately, there isn't a way to change the signature of virtual methods in a backwards compatible way.

One option is adding a new virtual method with the new signature, and then using `GDVIRTUAL_IS_OVERRIDDEN()` to call the new one if implemented, and then falling back on the old one if it isn't implemented?

I don't think we've actually done that anywhere yet - so far, I think we've always found some other way to work around it, so I don't have any examples to point to. However, we have a similar issue with GDExtension interface functions, and there we have the convention of adding an incrementing integer to each new version of the function. Not the most attractive API, but it works...

darksylinc commented:
All checks passed. All questions cleared. This PR is ready for review.

bruvzg commented on core/io/file_access.cpp:
```diff
@@ -985,6 +997,11 @@ void FileAccess::_bind_methods() {
+	BIND_ENUM_CONSTANT(SAVE_INTEGRITY_DEFAULT);
+	BIND_ENUM_CONSTANT(SAVE_INTEGRITY_NONE);
+	BIND_ENUM_CONSTANT(SAVE_INTEGRITY_SAVE_SWAP);
+	BIND_ENUM_CONSTANT(SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC);
```
```suggestion
	BIND_ENUM_CONSTANT(SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC);
	BIND_ENUM_CONSTANT(SAVE_INTEGRITY_MAX);
```

bruvzg commented on core/io/file_access.h:
```diff
@@ -84,6 +84,15 @@ class FileAccess : public RefCounted {
+		SAVE_INTEGRITY_DEFAULT,
+		SAVE_INTEGRITY_NONE,
+		SAVE_INTEGRITY_SAVE_SWAP,
+		SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC,
```
```suggestion
		SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC,
		SAVE_INTEGRITY_MAX,
```

bruvzg commented on doc/classes/OS.xml:
```diff
@@ -764,6 +764,16 @@
+				[b]Warning:[/b] Setting the default to [constant FileAccess.SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC] can cause a severe performance degradation. It's not recommended to set the default to this value.
+				[b]Note:[/b] On Windows, a value other than [constant FileAccess.SAVE_INTEGRITY_NONE] may interact negatively with certain antivirus programs. In this case, you'll have to set it to [constant FileAccess.SAVE_INTEGRITY_NONE] to prevent file locking issues.
+			</description>
+		</method>
```
```diff
-<method name="set_use_file_access_save_and_swap">
+<method name="set_use_file_access_save_and_swap" deprecated="Use [method set_default_save_integrity_level] instead.">

bruvzg commented on drivers/unix/file_access_unix.cpp:
```diff
@@ -62,9 +62,11 @@ void FileAccessUnix::check_errors(bool p_write) const {
 	_close();
 
+	DEV_ASSERT(p_integrity_level != SAVE_INTEGRITY_DEFAULT); // It should have already been resolved.
+
```
```suggestion
	ERR_FAIL_INDEX_V(p_integrity_level, SAVE_INTEGRITY_MAX, ERR_INVALID_PARAMETER);
```

bruvzg commented on drivers/windows/file_access_windows.cpp:
```diff
@@ -97,7 +97,7 @@ String FileAccessWindows::fix_path(const String &p_path) const {
 }
 
-Error FileAccessWindows::open_internal(const String &p_path, int p_mode_flags) {
+Error FileAccessWindows::open_internal(const String &p_path, int p_mode_flags, SaveIntegrityLevel p_integrity_level) {
```
```suggestion
Error FileAccessWindows::open_internal(const String &p_path, int p_mode_flags, SaveIntegrityLevel p_integrity_level) {

	DEV_ASSERT(p_integrity_level != SAVE_INTEGRITY_DEFAULT); // It should have already been resolved.
	ERR_FAIL_INDEX_V(p_integrity_level, SAVE_INTEGRITY_MAX, ERR_INVALID_PARAMETER);
```

bruvzg commented on core/crypto/crypto.cpp:
```diff
@@ -228,7 +228,7 @@ String ResourceFormatLoaderCrypto::get_resource_type(const String &p_path) const
 }
 
-Error ResourceFormatSaverCrypto::save(const Ref<Resource> &p_resource, const String &p_path, uint32_t p_flags) {
+Error ResourceFormatSaverCrypto::save(const Ref<Resource> &p_resource, const String &p_path, uint32_t p_flags, const int p_integrity_level) {
```
```suggestion
Error ResourceFormatSaverCrypto::save(const Ref<Resource> &p_resource, const String &p_path, uint32_t p_flags, int p_integrity_level) {
```

Usually we do not use `const` for primitive types, also any reason it's not using `FileAccess::SaveIntegrityLevel p_integrity_level)` directly?

darksylinc replied:
Ah! I missed this one. Removed. Thanks.

bruvzg commented on core/io/image.h:
```diff
@@ -333,13 +342,13 @@ class Image : public Resource {
 	static Ref<Image> load_from_file(const String &p_path);
-	Error save_png(const String &p_path) const;
-	Error save_jpg(const String &p_path, float p_quality = 0.75) const;
+	Error save_png(const String &p_path, int p_integrity_level = 0) const;
```
Same here, why not `FileAccess::SaveIntegrityLevel`?

darksylinc replied:
The only reason I'm using `int` instead of `FileAccess::SaveIntegrityLevel` is to avoid another `#include` as I need to bring file_access.h in. I've seen that pattern in other parts of the API, but used inconsistently.

Personally, I prefer strong type safety; and the fact that header inclusion is exploding throughout the database should be addressed separately(\*).

I can easily change it to `FileAccess::SaveIntegrityLevel`.
Thoughts?


(\*) Something needs to be done about that. A single header is changed and between 25-75% of the project needs to be built again. It's really bad for productivity.

bruvzg replied:
For unexposed `ResourceFormatSaver` it's OK as is.

But I would change it for public `Image` methods, some parts of the code avoid includes from other parts of the complex dependencies, but both `Image` and `FileAccess` are part `core/io`, so adding extra include should be fine.

darksylinc replied:
`ResourceFormatSaver` is part of `core/io` too, so if that's the criteria we can add the header there too.

On the other hand, ResourceFormatSaver is also included in a lot of other modules headers, while file_access.h isn't.

darksylinc replied:
I stand corrected. Image is included in a lot of other module headers too. It makes sense for `ResourceFormatSaver` to also become typed.

bruvzg commented on core/io/resource_saver.h:
```diff
@@ -82,9 +82,12 @@ class ResourceSaver {
 		FLAG_SAVE_BIG_ENDIAN = 16,
 		FLAG_COMPRESS = 32,
 		FLAG_REPLACE_SUBRESOURCE_PATHS = 64,
+		FLAG_SAVE_INTEGRITY_NONE = 128,
+		FLAG_SAVE_INTEGRITY_SAVE_SWAP = 256,
+		FLAG_SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC = FLAG_SAVE_INTEGRITY_NONE | FLAG_SAVE_INTEGRITY_SAVE_SWAP,
```
Are these flags used anywhere?

darksylinc replied:
Nooo!!!!

This is a left-over from a previous, scrapped idea. Well spotted! Removing this as well.

bruvzg commented on core/core_bind.cpp:
```diff
@@ -239,7 +239,12 @@ void OS::close_midi_inputs() {
 }
 
 void OS::set_use_file_access_save_and_swap(bool p_enable) {
```
```suggestion
#ifndef DISABLE_DEPRECATED
void OS::set_use_file_access_save_and_swap(bool p_enable) {
```

bruvzg commented on core/core_bind.cpp:
```diff
@@ -239,7 +239,12 @@ void OS::close_midi_inputs() {
-	FileAccess::set_backup_save(p_enable);
+	WARN_DEPRECATED_MSG("Use set_default_save_integrity_level() instead.");
+	FileAccess::set_default_save_integrity_level(p_enable ? FileAccess::SAVE_INTEGRITY_SAVE_SWAP : FileAccess::SAVE_INTEGRITY_NONE);
+}
```
```suggestion
}
#endif
```

bruvzg commented on core/core_bind.h:
```diff
@@ -228,6 +228,7 @@ class OS : public Object {
 	Key find_keycode_from_string(const String &p_code) const;
 
 	void set_use_file_access_save_and_swap(bool p_enable);
```
```suggestion
#ifndef DISABLE_DEPRECATED
	void set_use_file_access_save_and_swap(bool p_enable);
#endif

```

bruvzg commented on core/core_bind.cpp:
```diff
@@ -723,6 +728,7 @@ void OS::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("find_keycode_from_string", "string"), &OS::find_keycode_from_string);
 
 	ClassDB::bind_method(D_METHOD("set_use_file_access_save_and_swap", "enabled"), &OS::set_use_file_access_save_and_swap);
```
```suggestion
#ifndef DISABLE_DEPRECATED
	ClassDB::bind_method(D_METHOD("set_use_file_access_save_and_swap", "enabled"), &OS::set_use_file_access_save_and_swap);
#endif
```

darksylinc committed:
Add save integrity levels.

This PR adds an argument to most save functions to specify what level of
integrity should Godot attempt to guarantee when writing a file.

This in preparation for PR #98361; which will introduce a full sync
barrier when performing save operations to avoid data loss.

This PR puts such expensive sync() behind the enum
SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC; which should be used sparingly for
important information.

This PR also sets most cache files to be saved as either
SAVE_INTEGRITY_SAVE_SWAP (old default behavior) or SAVE_INTEGRITY_NONE;
because Godot can easily generate thousands of these file saving
operations. Performing thousands of full syncs() would be too expensive,
specially when these files can be regenerated if they're corrupted or
lost.

The PR is incomplete as it does not yet expose integrity levels to the
user, which is relevant for certain file saving operations, such as when
game save files that store the player's progress (those should be done
with SAVE_INTEGRITY_SAVE_SWAP_PLUS_SYNC).

dalexeev commented on core/io/config_file.h:
```diff
@@ -61,7 +67,7 @@ class ConfigFile : public RefCounted {
 	void erase_section_key(const String &p_section, const String &p_key);
 
-	Error save(const String &p_path);
+	Error save(const String &p_path, FileAccess::SaveIntegrityLevel p_integrity_level = FileAccess::SAVE_INTEGRITY_DEFAULT);
```
There are also `save_encrypted()` and `save_encrypted_pass()`.

darksylinc replied:
Done.

MarianoGnu commented on core/io/resource_saver.cpp:
```diff
@@ -41,9 +41,17 @@ bool ResourceSaver::timestamp_on_save = false;
+
+#ifndef DISABLE_DEPRECATED
+	// If not, we fall back on the old one. Notice the "_save_compat_100447" alias we're using here instead of "_save".
+	GDVIRTUAL_CALL(_save_compat_100447, p_resource, p_path, p_flags, err);
```
what is going on here? `_save_compat_100447` [only takes a `path` argument](https://github.com/godotengine/godot/pull/100447/files#diff-3b9dd8590c4e9192c964dc726efa3f1a3911d6956a7902fa19e90453f8cef9edR52) but i see 4 arguments here, including a non reference or pointer `err` argument.
How does this work?

darksylinc replied:
You were looking at `ConfigFile::_save_compat_100447`; but you should be looking at `ResourceFormatSaver::_save_compat_100447` which takes a `path` argument and a `flags` argument. 2 arguments in total.

There are other issues with `ResourceFormatSaver::_save_compat_100447`, I'll be pushing a fix in 10 or 20 minutes.

darksylinc replied:
OK all issues I could find are fixed. CI is till running though.

darksylinc commented:
I've rebased this PR to latest master. It's ready for review and merge.

Unlike last time, now we have the infrastructure for backwards compatibility with GD script virtual calls.

darksylinc commented on core/io/file_access.h:
```diff
@@ -141,7 +152,7 @@ class FileAccess : public RefCounted {
 	}
 
-	static Ref<FileAccess> _open(const String &p_path, ModeFlags p_mode_flags);
+	static Ref<FileAccess> _open(const String &p_path, ModeFlags p_mode_flags, SaveIntegrityLevel p_integrity_level = FileAccess::SAVE_INTEGRITY_DEFAULT);
```
I'm willing to accept either way.

My reason for preferring `FileAccess::SAVE_INTEGRITY_DEFAULT` is that it makes it easier to copy paste: any API user will write `foo._open( ... )` and wants to override the default, will see `SaveIntegrityLevel p_integrity_level = SAVE_INTEGRITY_DEFAULT`.

Copy paste SAVE_INTEGRITY_DEFAULT, then:

1. Compiler error.
2. Autocomplete doesn't work either, because `SAVE_INTEGRITY_` does not resolve either.
3. Once the user realizes what's happening, he needs to go to file_access.h, hit Go To Definition on `SAVE_INTEGRITY_DEFAULT` then scroll to the top and see it belongs to `FileAccess`. Aha! Write `FileAccess::`

This is solved by placing the member on the declaration.

Copy pasting `FileAccess::SAVE_INTEGRITY_DEFAULT` works out of the box every time, and `FileAccess::SAVE_INTEGRITY_` works with autocomplete. The quality of life for other C++ users who consume the API is much improved.

AThousandShips commented on core/io/file_access.compat.inc:
```diff
@@ -91,6 +95,7 @@ void FileAccess::store_pascal_string_bind_compat_78289(const String &p_string) {
 }
 
 void FileAccess::_bind_compatibility_methods() {
+	ClassDB::bind_compatibility_static_method("FileAccess", D_METHOD("open", "path", "flags"), &FileAccess::_open_bind_compat_100447);
```
Compatibility methods should be added at the end regardless of other order

darksylinc replied:
Ok thanks! Done.
